### PR TITLE
MempoolP2PTester script

### PR DESCRIPTION
**Pull request by emmanueldenloye**

### CODE REVIEW

File: cabal.project [hunk 1]
```diff
@@ -6,12 +6,16 @@ package aeson
...
+    tag: 01d2f73475b1fe4bdc67e615b74055b2db8fa3a2
 
 source-repository-package
     type: git
     location: https://github.com/kadena-io/thyme.git
     tag: 6ee9fcb026ebdb49b810802a981d166680d867c9
+source-repository-package
+    type: git
+    location: https://github.com/LeventErkok/sbv.git
+    tag: a8c0375bd096aa46209fd8f9f9d9168f932d07dd
```
Comment (larskuhtz) [inline cabal.project]: Why can't we use latest from Hackage? I think it works fine with pact.

Why is the dependency on SBV needed at all? Can we build pact with `-f-build-tools` and get rid of this dependency.?

File: lib/TXG/Utils.hs [hunk 2]
```diff
@@ -259,6 +260,9 @@ chainBasePath :: ChainwebVersion -> ChainId -> T.Text
 chainBasePath v cid = "chainweb/0.0/" <> chainwebVersionToText v
     <> "/chain/" <> cidToText cid
 
+chainBasePath' :: ChainwebVersion -> T.Text
+chainBasePath' v = "chainweb/0.0/" <> chainwebVersionToText v
```
Comment (larskuhtz) [inline lib/TXG/Utils.hs]: Maybe just ball it `basPath` because it doesn't include the chain?

File: exec/MPT/Utils.hs [hunk 3]
```diff
@@ -0,0 +1,3 @@
+-- |
+
+module MPT.Utils where
```
Comment (larskuhtz) [inline exec/MPT/Utils.hs]: What is the purpose of this?

File: exec/TXG/Types.hs [hunk 4]
```diff
@@ -155,6 +163,22 @@ listenkeys = ListenerRequestKey . T.decodeUtf8
 -- Args
 -------
 
+read' :: Read a => String -> ReadM a
+read' msg = eitherReader (bimap (const msg) id . readEither)
```
Comment (larskuhtz) [inline exec/TXG/Types.hs]: The same definition shows up in exec/MPT/Types.hs. Maybe export from here or move to exec/MPT/Utils.hs?

File: exec/TXG/Types.hs [hunk 5]
```diff
@@ -155,6 +163,22 @@ listenkeys = ListenerRequestKey . T.decodeUtf8
...
+
+pHostAddress' :: Maybe String -> O.Parser HostAddress
+pHostAddress' service = HostAddress <$> pHostname service <*> pPort service
```
Comment (larskuhtz) [inline exec/TXG/Types.hs]: The same definition shows up in exec/MPT/Types.hs. Maybe export from here move to exec/MPT/Utils.hs?

File: exec/TXG/Types.hs [hunk 6]
```diff
@@ -298,17 +361,37 @@ data TXGConfig = TXGConfig
...
+  -> BatchSize
+  -> Verbose
+  -> GasLimit
+  -> GasPrice
+  -> TTLSeconds
+  -> HostAddress
+  -> IO TXGConfig
+mkConfig mdistribution version batchSize v gl gp ttl hostAddr =
   TXGConfig mdistribution mempty
   <$> pure hostAddr
```
Comment (larskuhtz) [inline exec/TXG/Types.hs]:
```suggestion
  TXGConfig mdistribution mempty hostAddr
```

File: exec/TXG/Types.hs [hunk 7]
```diff
@@ -298,17 +361,37 @@ data TXGConfig = TXGConfig
...
-  <*> pure (verbose config)
-  <*> pure (gasLimit config)
-  <*> pure (gasPrice config)
-  <*> pure (timetolive config)
+  <*> pure version
+  <*> pure batchSize
+  <*> pure v
+  <*> pure gl
+  <*> pure gp
+  <*> pure ttl
```
Comment (larskuhtz) [inline exec/TXG/Types.hs]: Long lists of positional parameters are a little hard to read and maintain. Since, there's only a single non-pure argument, why not use record initialization syntax with explicitly name record fields? What about doing something like:

```haskell
mgr <- unsafeManager
return $ TXGConfig
   { ...
   }
```

File: exec/TXG/Types.hs [hunk 8]
```diff
@@ -285,6 +339,15 @@ data TXGState = TXGState
...
   } deriving (Generic)
 
+type Cut = LB.ByteString
+
+data MPTState = MPTState
+  { mptGen    :: !(Gen (PrimState IO))
+  , mptCounter :: !(TVar TXCount)
+  , mptLatestCut :: !(TVar Cut)
+  , mptChains :: !(NESeq ChainId)
+  } deriving (Generic)
```
Comment (larskuhtz) [inline exec/TXG/Types.hs]: I would expect this definition in exec/MPT/Types.hs. Is there a reason why it is defined here?

File: exec/TXG/Types.hs (hunk 7 above)
Comment (larskuhtz) [reply exec/TXG/Types.hs]: I am not sure how useful this function is generally. The only thing it does is calling `unsafeManager` at the cost of hiding the constructor field names from the record constructor.

File: exec/MPT.hs [hunk 9]
```diff
@@ -0,0 +1,573 @@
...
+import           Data.Aeson.Lens
+import qualified Data.ByteString.Lazy as LB
+import           Data.Generics.Product.Fields (field)
+import           Data.Foldable
+import           Data.Functor.Compose
+import qualified Data.HashMap.Strict as HM
+import           Data.Int
+import qualified Data.List.NonEmpty as NEL
+import           Data.Map (Map)
+import qualified Data.Map as M
```
Comment (larskuhtz) [inline exec/MPT.hs:50]: Is there a reason not to use `Data.Map.Strict`?

File: exec/MPT.hs [hunk 10]
```diff
@@ -0,0 +1,573 @@
...
+worker config = do
+    tv  <- newTVarIO 0
+    mgr <- unsafeManager
+    res <- runExceptT $ do
+      ni <- ExceptT $ getNodeInfo mgr (head $ mpt_hostAddresses config)
+      cut <- withExceptT show $ ExceptT $ queryCut mgr (head $ mpt_hostAddresses config) ni
+      return (cut,ni)
+    (cut,ni) <- case res of
+          Left err -> die err
+          Right (cut,ni) -> pure (cut, ni)
```
Comment (larskuhtz) [inline exec/MPT.hs:113]:
```suggestion
    (cut,ni) <- either die pure res
```

File: exec/MPT.hs [hunk 11]
```diff
@@ -0,0 +1,573 @@
...
+mainInfo :: ProgramInfo MPTArgs
+mainInfo =
+  programInfo
+  "MempoolP2PTester"
+  mpt_scriptConfigParser
+  defaultMPTArgs
+
+worker :: MPTArgs -> IO ()
+worker config = do
+    tv  <- newTVarIO 0
```
Comment (larskuhtz) [inline exec/MPT.hs:105]: If this is just a monotonically increasing concurrent counter, you may consider using just an `IORef` along with `atomicModifyIORef'`, which has less overhead than an STM tx.

File: exec/MPT.hs [hunk 12]
```diff
@@ -0,0 +1,573 @@
...
+    withLog inner = Y.withHandleBackend_ id (logconfig ^. Y.logConfigBackend)
+        $ \backend -> Y.withLogger (logconfig ^. Y.logConfigLogger) backend inner
+    act :: TVar TXCount -> TVar Cut -> HostAddress -> Manager -> LoggerT T.Text IO ()
+    act tv tcut host@(HostAddress h p) mgr = localScope (const [(hostnameToText h, portToText p)]) $
+      coinTransfers config mgr tv tcut host defaultTimingDist
+
+
+cutLoop :: Manager -> TVar Cut -> HostAddress -> NodeInfo -> IO ()
+cutLoop mgr tcut addr ni = forever $ do
+  threadDelay $ 60 * 1000000
```
Comment (larskuhtz) [inline exec/MPT.hs:130]:
```suggestion
  threadDelay 60_000_000
```

File: exec/MPT.hs [hunk 13]
```diff
@@ -0,0 +1,573 @@
...
+  ecut <- queryCut mgr addr ni
+  cut <- case ecut of
+    Left err -> die $ show err
+    Right cut -> pure cut
```
Comment (larskuhtz) [inline exec/MPT.hs:134]:
```suggestion
  cut <- either (die . show) pure ecut
```

File: exec/MPT.hs [hunk 14]
```diff
@@ -0,0 +1,573 @@
...
+  atomically $
+    -- ASSUMPTION --
+    -- Given a delay of 60 seconds, let's assume that the cut has changed.
+    -- ASSUMPTION --
```
Comment (larskuhtz) [inline exec/MPT.hs:138]: Why this assumption needed? What would be the consequences if does not hold?

File: exec/MPT.hs [hunk 15]
```diff
@@ -0,0 +1,573 @@
...
+        case ecids of
+          Left ex -> fail (displayException ex)
+          Right cids -> return (S.fromList cids)
+
+queryCut :: Manager -> HostAddress -> NodeInfo -> IO (Either ApiError Cut)
+queryCut mgr h ni = do
+  let v = _nodeInfo_chainwebVer ni
+      url = "https://" <> hostAddressToText h <> "/chainweb/0.0/" <> v <> "/cut"
+  req <- parseRequest $ T.unpack url
+  res <- handleRequest req mgr
```
Comment (larskuhtz) [inline exec/MPT.hs:218]: Maybe wrap that in retry?

File: exec/MPT.hs [hunk 16]
```diff
@@ -0,0 +1,573 @@
...
+  pure $ responseBody <$> res
+
+-- TODO Make this a configuration paramteter
+confirmationDepth :: Int
+confirmationDepth = 6
+
+loopUntilConfirmationDepth :: Int -> ChainId -> BlockHeight -> TVar Cut -> IO Int64
+loopUntilConfirmationDepth seconds cid startHeight tcut = do
+  threadDelay $ seconds * 1000000
```
Comment (larskuhtz) [inline exec/MPT.hs]:
```suggestion
  threadDelay $ seconds * 1_000_000
```

File: exec/MPT.hs [hunk 17]
```diff
@@ -0,0 +1,573 @@
...
+    logconfig = Y.defaultLogConfig
+        & Y.logConfigLogger . Y.loggerConfigThreshold .~ Info
+    withLog inner = Y.withHandleBackend_ id (logconfig ^. Y.logConfigBackend)
+        $ \backend -> Y.withLogger (logconfig ^. Y.logConfigLogger) backend inner
+    act :: TVar TXCount -> TVar Cut -> HostAddress -> Manager -> LoggerT T.Text IO ()
+    act tv tcut host@(HostAddress h p) mgr = localScope (const [(hostnameToText h, portToText p)]) $
+      coinTransfers config mgr tv tcut host defaultTimingDist
+
+
+cutLoop :: Manager -> TVar Cut -> HostAddress -> NodeInfo -> IO ()
```
Comment (larskuhtz) [inline exec/MPT.hs:128]: An alternative, push based, approach for getting notifications on new blocks is to subscribe to the header update stream. But polling the cut is fine, too.

File: exec/MPT.hs [hunk 18]
```diff
@@ -0,0 +1,573 @@
...
+worker :: MPTArgs -> IO ()
+worker config = do
+    tv  <- newTVarIO 0
+    mgr <- unsafeManager
+    res <- runExceptT $ do
+      ni <- ExceptT $ getNodeInfo mgr (head $ mpt_hostAddresses config)
+      cut <- withExceptT show $ ExceptT $ queryCut mgr (head $ mpt_hostAddresses config) ni
+      return (cut,ni)
+    (cut,ni) <- case res of
+          Left err -> die err
```
Comment (larskuhtz) [inline exec/MPT.hs:112]: Maybe do some retries before dying? Could be done in `queryCut`.

File: exec/MPT.hs [hunk 19]
```diff
@@ -0,0 +1,573 @@
...
+  pure $ responseBody <$> res
+
+-- TODO Make this a configuration paramteter
+confirmationDepth :: Int
+confirmationDepth = 6
+
+loopUntilConfirmationDepth :: Int -> ChainId -> BlockHeight -> TVar Cut -> IO Int64
+loopUntilConfirmationDepth seconds cid startHeight tcut = do
+  threadDelay $ seconds * 1000000
+  cut <- atomically $ readTVar tcut
```
Comment (larskuhtz) [inline exec/MPT.hs]: `readTVarIO` more efficient than `atomically . readTVar`. Internally, it's just a pointer dereference.

```suggestion
  cut <- readTVarIO tcut
```

File: exec/MPT.hs [hunk 20]
```diff
@@ -0,0 +1,573 @@
...
+  let height = fromIntegral $ fromJuste $ cut ^? key "hashes" . key (cidToText cid) . key "height" . _Integer
+  if height < (startHeight + confirmationDepth)
+    then do
+      loopUntilConfirmationDepth seconds cid startHeight tcut
+    else getCurrentTimeInt64
```
Comment (larskuhtz) [inline exec/MPT.hs]: Given that `tcut` is already a `TVar`, you may want to await a change instead of polling. The following will `retry` each time when `tcut` is updated until the condition is satisfied. 

```suggestion
  atomically $ do
    cut <- readTVar tcut
    let height = fromIntegral $ fromJuste $ cut ^? key "hashes" . key (cidToText cid) . key "height" . _Integer
    if height < (startHeight + confirmationDepth)
      then retry
      else return ()
  getCurrentTimeInt64    
```
 
 If you stick to polling, you may consider just using `IORef` and update it with `atomicModifyIORef`.

File: exec/MPT.hs [hunk 21]
```diff
@@ -0,0 +1,573 @@
...
+
+
+check' :: RetryStatus -> Either String a -> IO Bool
+check' _ ev = pure $
+  case ev of
+    Left _retry -> True
+    Right _result -> False
```
Comment (larskuhtz) [inline exec/MPT.hs]: Is this work in progress? Otherwise, why is it in `IO` and not just `check' _ = isList`?

Comment (larskuhtz) [reply exec/MPT.hs]: Also, is `Left` supposed to be `True`?

File: exec/MPT.hs [hunk 22]
```diff
@@ -0,0 +1,573 @@
...
+
+loop
+  :: (MonadIO m, MonadLog T.Text m)
+  => TVar Cut
+  -> FilePath
+  -> TXG MPTState m (Sim.ChainId, NEL.NonEmpty (Maybe Text), NEL.NonEmpty (Command Text))
+  -> TXG MPTState m ()
+loop tcut dbFile f = forever $ do
+  liftIO $ threadDelay $ 1_000_0000 * 60
```
Comment (larskuhtz) [inline exec/MPT.hs]:
```suggestion
  liftIO $ threadDelay 60_000_0000
```

File: exec/MPT.hs [hunk 23]
```diff
@@ -0,0 +1,573 @@
...
+      countTV <- gets mptCounter
+      batch <- asks confBatchSize
+      liftIO . atomically $ modifyTVar' countTV (+ fromIntegral batch)
+      count <- liftIO $ readTVarIO countTV
+      lift . logg Info $ "Transaction count: " <> T.pack (show count)
+      lift . logg Info $ "Transaction requestKey: " <> T.pack (show rks)
+      let retrier = retrying policy check' . const
+          policy :: RetryPolicyM IO
+          policy = exponentialBackoff 250_000 <> limitRetries 3
+      liftIO $ withAsync (retrier $ pollRequestKeys config cid rks) $ \a -> do
```
Comment (larskuhtz) [inline exec/MPT.hs]: Why is `(retrier $ pollRequestKeys config cid rks)` scheduled asynchronously? It seems that it is awaited immediately, so the execution on the calling thread blocks until the `pollRequestKeys` thread has finished.

File: exec/MPT.hs [hunk 24]
```diff
@@ -0,0 +1,573 @@
...
+        wait a >>= \case
+          Left err -> liftIO $ putStrLn $ printf "[ERROR] Caught this error while polling for these request keys (%s) %s" (show rks) err
```
Comment (larskuhtz) [inline exec/MPT.hs]: Why not using logger like elsewhere?

File: .github/workflows/ci.yaml [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+    strategy:
+      fail-fast: false
+      matrix:
+        ghc: ['8.8.4']
+        cabal: ['3.4']
+        os: ['ubuntu-18.04', 'ubuntu-20.04', 'macOS-latest']
+
+    steps:
+    - name: 'GitHub actions env workaround'
+      run: 'echo "ACTIONS_ALLOW_UNSECURE_COMMANDS=true" >> $GITHUB_ENV'
```
Comment (larskuhtz) [inline .github/workflows/ci.yaml:18]: Why is this needed?

File: exec/TXG/Types.hs (hunk 5 above)
Comment (emmanueldenloye) [reply exec/TXG/Types.hs]: I removed this entirely by building `hostaddress` correctly.

File: exec/MPT/Utils.hs (hunk 3 above)
Comment (emmanueldenloye) [reply exec/MPT/Utils.hs]: I've removed this module.

File: exec/TXG/Types.hs (hunk 4 above)
Comment (emmanueldenloye) [reply exec/TXG/Types.hs]: Resolved.

File: exec/TXG/Types.hs (hunk 8 above)
Comment (emmanueldenloye) [reply exec/TXG/Types.hs]: This is a good point. I think I'll move it.
